### DISCUSSION TITLE
html-pdf API route returns NO_RESPONSE_FROM_FUNCTION Error: write EPIPE

### DISCUSSION THREAD

**Discussion by eivindml** [vercel/vercel#4460] [Help]
I'm trying to create a small API function that receives a html string, and returns a PDF.

I have this small serverless function:

```js
import { NextApiRequest, NextApiResponse } from "next";
import pdf from "html-pdf";

export default (req: NextApiRequest, res: NextApiResponse) => {
  if (req.method === "POST") {
    pdf.create(req.body).toBuffer(function(_, buffer) {
      res.send(buffer);
    });
  } else {
    res.status(400).send("Should be POST.");
  }
};
```

This works great locally, but I can't get it to work when deploying to Vercel. I get a `502` error and this message when testing the API:

```
An error occurred with this application.

NO_RESPONSE_FROM_FUNCTION
```

Here is a deployed version of it: https://generator-test-theta.now.sh/api/generate

Why is this happening? How can I debug this? Or is there some limitations that makes this not possible to do?

**Comment by mcsdevv** [comment]
Hello @eivindml, the best way to debug this is by looking at the realtime request logs - there you will find the error being thrown.

**Reply by eivindml** [reply]
Thanks! Where do I find these?

**Reply by mcsdevv** [reply]
From your Vercel dashboard, select your project and click "View Function Logs".

**Comment by eivindml** [comment]
This is the error I'm getting. I'm not sure what the exact problem I'm with regards the the Vercel serverless function. It works locally, and this error doesn't make me that much more informed. 

Any tips?

```bash
[POST] /api/generate
08:55:23:54
2020-05-25T06:55:23.713Z	450e8c63-a7a9-4819-bab9-913315205f84	ERROR	Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:154:25)
    at writeGeneric (internal/stream_base_commons.js:145:3)
    at Socket._writeGeneric (net.js:784:11)
    at Socket._write (net.js:796:8)
    at doWrite (_stream_writable.js:442:12)
    at writeOrBuffer (_stream_writable.js:426:5)
    at Socket.Writable.write (_stream_writable.js:317:11)
    at PDF.PdfExec [as exec] (/var/task/node_modules/html-pdf/lib/pdf.js:141:15)
    at PDF.PdfToBuffer [as toBuffer] (/var/task/node_modules/html-pdf/lib/pdf.js:44:8)
    at module.exports.rvJR.__webpack_exports__.default (/var/task/.next/serverless/pages/api/generate.js:277:70) {
  errno: 'EPIPE',
  code: 'EPIPE',
  syscall: 'write'
}
2020-05-25T06:55:23.716Z	450e8c63-a7a9-4819-bab9-913315205f84	ERROR	Uncaught Exception 	{"errorType":"TypeError","errorMessage":"The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received undefined","code":"ERR_INVALID_ARG_TYPE","stack":["TypeError [ERR_INVALID_ARG_TYPE]: The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received undefined","    at Function.byteLength (buffer.js:723:11)","    at sendData (/var/task/node_modules/next/dist/next-server/server/api-utils.js:30:205)","    at ServerResponse.apiRes.send (/var/task/node_modules/next/dist/next-server/server/api-utils.js:4:236)","    at /var/task/.next/serverless/pages/api/generate.js:278:11","    at execPdfToBuffer (/var/task/node_modules/html-pdf/lib/pdf.js:45:21)","    at respond (/var/task/node_modules/html-pdf/lib/pdf.js:126:14)","    at ChildProcess.onError (/var/task/node_modules/html-pdf/lib/pdf.js:134:46)","    at ChildProcess.emit (events.js:310:20)","    at Process.ChildProcess._handle.onexit (internal/child_process.js:273:12)","    at onErrorNT (internal/child_process.js:469:16)"]}
Unknown application error occurred
```

**Comment by eivindml** [comment]
Any help here? I would like to understand how Vercel works differently so this doesn't work, so that maybe I can fix this. 😄

**Reply by wwwflorencio** [reply]
Your function is probably trying to write the file on disk, verify how `html-pdf` lib works. You can write files just on /temp folder using now.sh.

Edit:
The folder is `/tmp` not `/temp`

**Reply by eivindml** [reply]
Ok, thank you! :) I will see if I maybe can configure where files are written.

**Reply by eivindml** [reply]
For anyone else seeing this, the folder is `/tmp` not `/temp`. But `html-pdf` uses this by default, so still didn't fix the issue. :( 

See here: https://www.npmjs.com/package/html-pdf#options

**Reply by wwwflorencio** [reply]
Sorry for the little typo with `tmp ` :D

Try to use `tmpdir` function to get the tmp path, this works for me in the past using vercel:

```javascript
const { tmpdir } = require('os')

const config = {
  directory: tmpdir()
}
```
```

**Reply by eivindml** [reply]
Tried this (see latest post), still not working. Starting to think this is something with how `pdf-html` works, which makes it incompatible with Vercel.

**Comment by styfle** [comment]
You can only write files to the `/tmp` directory. The rest of the file system is read only.

But I would suggest avoiding File IO altogether and instead write the data to the http response.

**Reply by eivindml** [reply]
Thank you for you reply!
I thought I avoided writing files by creating and sending a buffer directly?

> But I would suggest avoiding File IO altogether and instead write the data to the http response.

I would love to do this. 😃 Do you have a pointer to how I might achieve this? I wouldn't know what to google for, so any hint is of help.

**Comment by styfle** [comment]
Perhaps the issue is reading `req.body` which might not the type you are expecting.

See my answer here to read the raw body: https://github.com/vercel/vercel/discussions/4524#discussioncomment-20184

You are consuming the error `_` argument instead of printing it. There is probably more information there if you do a `console.log(_)`.

**Comment by eivindml** [comment]
This is the code I use now:

```javascript
import { NextApiRequest, NextApiResponse } from "next";
import pdf from "html-pdf";
import { tmpdir } from "os";

const options = {
  directory: tmpdir(),
  phantomPath: "./node_modules/phantomjs-prebuilt/bin/phantomjs"
};

export default (_: NextApiRequest, response: NextApiResponse) => {
  pdf.create("<p>Hello world!</p>", options).toBuffer(function(error, buffer) {
    if (error) {
      response.status(400).send(error.message);
    } else {
      response.send(buffer);
    }
  });
};
```

I removed the dependency on `req.body` to rule out this possible issue pointed out by @styfle . 

This is the error response from the API:

<img width="1153" alt="Skjermbilde 2020-06-01 kl  11 29 27" src="https://user-images.githubusercontent.com/202450/83395840-46502e00-a3fb-11ea-98a0-3df39d40f3bf.png">

The function log on vercel doesn't show anything special.

<img width="1036" alt="Skjermbilde 2020-06-01 kl  11 29 48" src="https://user-images.githubusercontent.com/202450/83395880-56680d80-a3fb-11ea-91cb-d8d9bb78cc9b.png">


I think the issue might be something specific with `html-pdf` that makes it give an error when running on the vercel server.

**Comment by styfle** [comment]
Looks like `phantomPath` binary in not included in the bundle.

Try using `path.resolve()` here:

```js
const options = {
  directory: tmpdir(),
  phantomPath: path.resolve("./node_modules/phantomjs-prebuilt/bin/phantomjs")
};
```

**Reply by eivindml** [reply]
Thanks, but unfortunately still same result. 😢

<img width="2325" alt="Skjermbilde 2020-06-02 kl  09 28 47" src="https://user-images.githubusercontent.com/202450/83492393-7eb34300-a4b3-11ea-8dd6-385a87d79a18.png">

**Comment by styfle** [comment]
Perhaps this will work:

```js
const options = {
  directory: tmpdir(),
  phantomPath: path.resolve("phantomjs-prebuilt/bin/phantomjs")
};
```

**Reply by eivindml** [reply]
Still no luck. 😔

I ended using `PDFKit` based on @rauchg's example:
https://twitter.com/rauchg/status/1067929593410019330 

That works great, even though it requires a bit more work than just rendering some html to a pdf. 

But I'm still curious too why this isn't working. So that I maybe can learn something after all this trial and errors. :P

**Reply by eivindml** [reply]
Reference for others that might end up here: you can see the source code for the function here: https://vercel.com/now-examples/nodejs-pdfkit/3grfykwfj/source

**Comment by ceduth** [comment]
This works on Vercel with node v18, and latest puppeteer-core. 
[full steps](https://github.com/marcbachmann/node-html-pdf/issues/586#issuecomment-1872992358).

```js
puppeteer.launch({
    args: [...chromium.args, '--hide-scrollbars', '--disable-web-security'],
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath(
      `https://github.com/Sparticuz/chromium/releases/download/v119.0.2/chromium-v119.0.2-pack.tar`
    ),
    headless: chromium.headless,
    ignoreHTTPSErrors: true,
});
```
